File: backend/app/schemas/application.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, field_validator

from app.models.application import APPLICATION_STATUSES
from app.schemas.job import JobRead


class ApplicationEventRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    event_type: str
    message: str
    created_at: datetime


class ApplicationCreate(BaseModel):
    job_id: uuid.UUID
    resume_id: uuid.UUID | None = None
    status: str = "discovered"

    @field_validator("status")
    @classmethod
    def valid_status(cls, v: str) -> str:
        if v not in APPLICATION_STATUSES:
            raise ValueError(f"status must be one of {APPLICATION_STATUSES}")
        return v


class ApplicationUpdate(BaseModel):
    status: str | None = None
    resume_id: uuid.UUID | None = None
    notes: str | None = None

    @field_validator("status")
    @classmethod
    def valid_status(cls, v: str | None) -> str | None:
        if v is not None and v not in APPLICATION_STATUSES:
            raise ValueError(f"status must be one of {APPLICATION_STATUSES}")
        return v


class NoteCreate(BaseModel):
    message: str


class ApplicationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    job_id: uuid.UUID
    job: JobRead
    resume_id: uuid.UUID | None
    status: str
    match_score: int | None
    notes: str | None
    discovered_at: datetime
    applied_at: datetime | None
    events: list[ApplicationEventRead]
    created_at: datetime

File: frontend/src/types/application.ts
import type { Job } from "./job";

export interface ApplicationEvent {
  id: string;
  event_type: string;
  message: string;
  created_at: string;
}

export interface Application {
  id: string;
  job_id: string;
  job: Job;
  resume_id: string | null;
  status: string;
  match_score: number | null;
  notes: string | null;
  discovered_at: string;
  applied_at: string | null;
  events: ApplicationEvent[];
  created_at: string;
}

export const APPLICATION_STATUSES = [
  "discovered",
  "matched",
  "saved",
  "preparing",
  "ready_for_review",
  "needs_manual_review",
  "submitted",
  "assessment",
  "interview",
  "offer",
  "rejected",
  "withdrawn",
] as const;
